src/main/utility/proc_maps.rs

```diff
@@ -0,0 +1,470 @@
+            begin: parse_field(caps.get(1).unwrap().as_str(), "begin", |s| {
+                usize::from_str_radix(s, 16)
+            })?,
+            end: parse_field(caps.get(2).unwrap().as_str(), "end", |s| {
+                usize::from_str_radix(s, 16)
+            })?,
+            read: {
+                let s = caps.get(3).unwrap().as_str();
+                match s {
+                    "r" => true,
+                    "-" => false,
+                    _ => return Err(format!("Couldn't parse read bit {}", s))?,
+                }
+            },
+            write: {
+                let s = caps.get(4).unwrap().as_str();
+                match s {
+                    "w" => true,
+                    "-" => false,
+                    _ => return Err(format!("Couldn't parse write bit {}", s))?,
+                }
+            },
+            execute: {
+                let s = caps.get(5).unwrap().as_str();
+                match s {
+                    "x" => true,
+                    "-" => false,
+                    _ => return Err(format!("Couldn't parse execute bit {}", s))?,
+                }
+            },
+            sharing: caps.get(6).unwrap().as_str().parse::<Sharing>()?,
+            offset: parse_field(caps.get(7).unwrap().as_str(), "offset", |s| {
+                usize::from_str_radix(s, 16)
+            })?,
+            device_major: parse_field(caps.get(8).unwrap().as_str(), "device_major", |s| {
+                i32::from_str_radix(s, 16)
+            })?,
+            device_minor: parse_field(caps.get(9).unwrap().as_str(), "device_minor", |s| {
+                i32::from_str_radix(s, 16)
+            })?,
+            // Undocumented whether this is actually base 10; change to 16 if we find
+            // counter-examples.
+            inode: parse_field(caps.get(10).unwrap().as_str(), "inode", |s| {
+                i32::from_str_radix(s, 10)
+            })?,
+            path: parse_field::<_, _, Box<dyn Error>>(
+                caps.get(11).unwrap().as_str(),
+                "path",
+                |s| match s {
+                    "" => Ok(None),
+                    s => Ok(Some(s.parse::<MappingPath>()?)),
+                },
+            )?,
+        })
+    }
+}
+
+/// Parses the contents of a /proc/[pid]/maps file
+pub fn parse_file_contents(mappings: &str) -> Result<Vec<Mapping>, Box<dyn Error>> {
+    let res: Result<Vec<_>, String> = mappings
+        .lines()
+        .map(|line| Mapping::from_str(line).map_err(|e| format!("Parsing line: {}\n{}", line, e)))
+        .collect();
+    Ok(res?)
+}
+
+/// Reads and parses the contents of a /proc/[pid]/maps file
+pub fn mappings_for_pid(pid: libc::pid_t) -> Result<Vec<Mapping>, Box<dyn Error>> {
+    use std::fs::File;
+    use std::io::Read;
+
+    let mut file = File::open(format!("/proc/{}/maps", pid))?;
+    let mut contents = String::new();
+    file.read_to_string(&mut contents)?;
+    Ok(parse_file_contents(&contents)?)
```

Comment:
This can be just `parse_file_contents(&contents)` I think.

Reply:
Done